File: 0017 Letter Combinations of a Phone Number/letter_combinations_of_a_phone_number_iterative.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        if not digits: # base case
            return res
        res.append("")
        my_dict = ['0', '1', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        for i in range(len(digits)):
            letters = my_dict[int(digits[i])]
            while len(res[0]) == i:
                prev = res.pop(0)
                for letter in letters:
                    res.append(prev + letter)
        return res
            

File: 17 Letter Combinations of a Phone Number/LetterCombinationsOfAPhoneNumber_Iterative.java
import java.lang.Character;
class LetterCombinationsOfAPhoneNumber_Iterative {
    public List<String> letterCombinations(String digits) {
        Queue<String> res = new LinkedList<>();
        if (digits == null || digits.length() == 0) { // base case
            return (List) res; // cast the queue to the return type
        }

        String[] map = {"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        res.add("");
        for (int i = 0; i < digits.length(); i++) {
            String s = map[Character.getNumericValue(digits.charAt(i))];
            while (res.peek().length() == i) {
                String prev = res.poll();
                for (int j = 0; j < s.length(); j++) {
                    res.add(prev + s.charAt(j));
                }
            }
        }
        return (List) res;
    }
}